underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java

```diff
@@ -387,35 +387,54 @@ public InputStream openAtPosition(String path, long pos) throws IOException {
   }
 
   @Override
-  public boolean rename(String src, String dst) throws IOException {
-    if (!isFile(src) && !isDirectory(src)) {
+  public boolean renameDirectory(String src, String dst) throws IOException {
+    if (!isDirectory(src)) {
       LOG.error("Unable to rename {} to {} because source does not exist.", src, dst);
       return false;
     }
     if (isFile(dst) || isDirectory(dst)) {
       LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
       return false;
     }
-    // Source exists and destination does not exist
-    if (isDirectory(src)) {
-      // Rename the source folder first
-      if (!copy(convertToFolderName(src), convertToFolderName(dst))) {
-        return false;
+    // Source exists and is a directory, and destination does not exist
+    // Rename the source folder first
+    if (!copy(convertToFolderName(src), convertToFolderName(dst))) {
+      return false;
+    }
+    // Rename each child in the src folder to destination/child
+    String [] children = list(src);
+    if (children == null) {
+      LOG.error("Failed to list path {}, aborting rename.", src);
+      return false;
+    }
+    for (String child : children) {
+      String childSrcPath = PathUtils.concatPath(src, child);
+      String childDstPath = PathUtils.concatPath(src, child);
+      boolean success;
+      if (isDirectory(childSrcPath)) {
+        // Recursive call
+        success = renameDirectory(childSrcPath, childDstPath);
```

Comment:
This call will redo the check to see if src is a directory, right? Is there a way to make that more efficient?